### ISSUE TITLE
Race condition in get-json test #1

### ISSUE THREAD

**Issue by Olaroll** [01-edu/public#939]
### The issue
There's a race condition that causes some working solutions to fail.

It happens because the test checks the `url` variable without giving the student's `getJSON()` function time to set it using `fetch()`

### The fix
My proposed solution is to add an `await` to the `getJSON()` call. 

Here's the relevant lines of code:
https://github.com/01-edu/public/blob/18d1db6e27430f77d819b1a9df1766427a6c2192/js/tests/get-json_test.js#L13-L19

**Comment by MSilva95** [comment]
Hello @Olaroll! I will take a look at the test, thank you!

**Comment by kigiri** [comment]
@Olaroll can you show the code you used as a solution, because the calling of the subsequent fetch and, generation of the URL is supposed to be synchronous, should not required to wait for the pending query.

**Comment by Olaroll** [comment]
Here's the code I passed with:
```js
async function getJSON(path, params) {
	const search = new URLSearchParams(params)
	const response = await fetch(path + '?' + search.toString())
	if (response.status !== 200)
		throw new Error(response.statusText)

	const {data, error} = await response.json()
	if (error)
		throw new Error(error)

	return data
}
```

I guess it's true that it can be written in a synchronous manner. It's a bit of a shame though, I think the await syntax looks cleaner.

Something like this is the intended way I guess?
```js
function getJSON(path, params) {
	const search = new URLSearchParams(params)
	return fetch(path + '?' + search.toString())
		.then(response => {
			if (response.status !== 200)
				throw new Error(response.statusText)
			return response.json()
		})
		.then(({data, error}) => {
			if (error)
				throw new Error(error)
			return data
		})
}
```

**Comment by kigiri** [comment]
both provided code pass the tests:
```js
const fakeFetch = async ({ data, error, ...opts } = {}) => ({
  ok: !opts.status,
  type: 'basic',
  status: 200,
  statusText: 'OK',
  json: async () => ({ data, error }),
  text: async () => JSON.stringify({ data, error }),
  ...opts,
})

let fetch = () => {}
async function getJSON(path, params) {
	const search = new URLSearchParams(params)
	const response = await fetch(path + '?' + search.toString())
	if (response.status !== 200)
		throw new Error(response.statusText)

	const {data, error} = await response.json()
	if (error)
		throw new Error(error)

	return data
}

let url 
fetch = async (arg) => fakeFetch({ url: (url = arg) }) 
const pending = getJSON('/test', { query: 'hello world', b: 5 }) 
console.log(url, '/test?query=hello+world&b=5')
```

`async` fonction are called synchronously, are you sure this was the code used to test @Olaroll ?

I'm unable to reproduce

**Comment by MSilva95** [comment]
@Olaroll  can you send the code that is failing and you say it is correct?
So that we can reproduce this problem

**Comment by Olaroll** [comment]
Hi, sorry for the late reply.

After doing some more digging into the problem, I've found out that my initial conclusion of the problem was wrong. The problem was caused by the use of [.replaceAll()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll), which is not supported by the testing server.

For some context: I was asked by another student to help debug their code. It was running fine on their browser, but didn't pass the test. I thought it looked like a race condition, because the `.replaceAll()` function **failed silently** on the test server. I was also unaware that async functions run synchronously up until the first await. That lead me to try and test my theory by delaying `fetch()` using `setTimeout()` in my own code. But that required using await and thus created a race condition where there wasn't one before. I got the same test result in both cases, so I had assumed my theory to be true.

Anyway, the main takeaway from this long winded explanation is that thrown errors from the `getJSON()` function are not shown to the student, which makes the error message quite meaningless.

I would suggest adding an error check for the promise returned from `getJSON()`, so that the student can get useful information in case their code fails, like was the case with `.replaceAll()`. I don't know whether a new issue should be opened for this or not, so feel free to close this issue if you feel a new one should be made.

The issue can be demonstrated with something like this:
```js
async function getJSON() {
	throw new Error('I am not shown to the student')
}
```

And testing the above code will give this error message from the system:
```
test #1 failed:
async ({ eq }) => {
  // check url parsing
  let url
  fetch = async (arg) => fakeFetch({ url: (url = arg) })
  const pending = getJSON('/test', { query: 'hello world', b: 5 })
  return eq(url, '/test?query=hello+world&b=5')
}

Error:
AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
+ actual - expected

+ undefined
- '/test?query=hello+world&b=5'
    at eq (file:///app/test.mjs:35:3)
    at get-json.js:22:10
    at runTests (file:///app/test.mjs:168:19)
```
